File: bindings/react-native/src/facets/workspace.ts
import UldrenLoom from '../NativeUldrenLoom';
import { authArgs, keyArgs } from '../internal';
import type { LoomAuth, LoomKey } from '../internal';

export function workspaceCreate(
  loomPath: string,
  name = '',
  facet = '',
  key?: LoomKey,
  auth?: LoomAuth
): Promise<string> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.workspaceCreate(loomPath, name, facet, passphrase, kek, authPrincipal, authPassphrase);
}

export function workspaceListJson(loomPath: string, key?: LoomKey, auth?: LoomAuth): Promise<string> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.workspaceListJson(loomPath, passphrase, kek, authPrincipal, authPassphrase);
}

export function workspaceRename(
  loomPath: string,
  workspace: string,
  newName: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.workspaceRename(loomPath, workspace, newName, passphrase, kek, authPrincipal, authPassphrase);
}

export function workspaceDelete(
  loomPath: string,
  workspace: string,
  key?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(key);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.workspaceDelete(loomPath, workspace, passphrase, kek, authPrincipal, authPassphrase);
}

File: bindings/android/src/commonMain/kotlin/ai/uldren/loom/facets/workspace.kt
package ai.uldren.loom

    /** Create a workspace and return its UUID string. */
expect fun Loom.workspaceCreate(
        path: String,
        name: String? = null,
        facet: String? = null,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): String


    /** List workspaces as JSON records with id, name, facets, and head. */
expect fun Loom.workspaceListJson(
        path: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): String


    /** Rename a workspace selected by UUID or current name. */
expect fun Loom.workspaceRename(
        path: String,
        workspace: String,
        newName: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** Delete a workspace selected by UUID or name. */
expect fun Loom.workspaceDelete(
        path: String,
        workspace: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )
